File: backend/voidswitch/api/admin/stats.py
"""Admin: dashboard summary statistics."""

from __future__ import annotations

from fastapi import APIRouter, Depends
from sqlalchemy import func, select, text
from sqlalchemy.ext.asyncio import AsyncSession

from voidswitch.constants import KeyStatus, NodeStatus
from voidswitch.core.auth import require_staff
from voidswitch.core.database import get_session
from voidswitch.models.db import ApiKey, Node, Provider, RequestLog, User, VoidToken
from voidswitch.models.schemas import StatsOut

router = APIRouter(prefix="/api/admin/stats", tags=["admin:stats"])


async def _count(session: AsyncSession, stmt) -> int:
    return int((await session.execute(stmt)).scalar_one() or 0)


def _since_24h(session: AsyncSession):
    """Database-native expression for "24 hours ago".

    Uses the DB clock so the comparison stays entirely server-side, avoiding
    Python↔DB timezone / type-mapping surprises that previously caused the
    24h window to be wider than intended (especially on SQLite)."""
    dialect = session.bind.dialect.name if session.bind is not None else "sqlite"
    if dialect == "postgresql":
        return func.now() - text("INTERVAL '24 hours'")
    return func.datetime("now", "-24 hours")


@router.get("", response_model=StatsOut)
async def stats(
    session: AsyncSession = Depends(get_session),
    _: User = Depends(require_staff),
) -> StatsOut:
    since = _since_24h(session)

    providers = await _count(session, select(func.count(Provider.id)))
    total_keys = await _count(session, select(func.count(ApiKey.id)))
    active_keys = await _count(
        session,
        select(func.count(ApiKey.id)).where(ApiKey.status == KeyStatus.ACTIVE.value),
    )
    total_proxies = await _count(session, select(func.count(Node.id)))
    active_proxies = await _count(
        session,
        # A node is "active" only when enabled AND marked active — the routing
        # and resurrector both require enabled, but the dashboard's disable
        # action leaves ``status`` untouched, so counting status alone would
        # over-report.
        select(func.count(Node.id)).where(
            Node.enabled.is_(True), Node.status == NodeStatus.ACTIVE.value
        ),
    )
    tokens = await _count(
        session, select(func.count(VoidToken.id)).where(VoidToken.deleted.is_(False))
    )

    requests_24h = await _count(
        session, select(func.count(RequestLog.id)).where(RequestLog.ts >= since)
    )
    success_24h = await _count(
        session,
        select(func.count(RequestLog.id)).where(
            RequestLog.ts >= since, RequestLog.success.is_(True)
        ),
    )
    tokens_24h = await _count(
        session,
        select(func.coalesce(func.sum(RequestLog.total_tokens), 0)).where(RequestLog.ts >= since),
    )
    avg_first_token_ms_24h = (
        await session.execute(
            select(func.avg(RequestLog.first_token_ms)).where(
                RequestLog.ts >= since,
                RequestLog.success.is_(True),
                RequestLog.stream.is_(True),
                RequestLog.first_token_ms.isnot(None),
            )
        )
    ).scalar_one()

    return StatsOut(
        providers=providers,
        active_keys=active_keys,
        total_keys=total_keys,
        active_proxies=active_proxies,
        total_proxies=total_proxies,
        tokens=tokens,
        requests_24h=requests_24h,
        success_24h=success_24h,
        failures_24h=requests_24h - success_24h,
        tokens_24h=tokens_24h,
        success_rate_24h=round(success_24h / requests_24h * 100, 1) if requests_24h else 0.0,
        avg_first_token_ms_24h=(
            round(avg_first_token_ms_24h, 1) if avg_first_token_ms_24h is not None else None
        ),
        avg_tokens_per_request_24h=(round(tokens_24h / requests_24h, 1) if requests_24h else 0.0),
    )

File: backend-go/internal/handlers/admin/stats.go
package admin

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/siiway/voidswitch/internal/database"
	"github.com/siiway/voidswitch/internal/models"
)

func getStats(c *gin.Context) {
	db := database.GetDatabase().DB
	user, ok := getCurrentUserGin(c)
	if !ok || !requireStaffGin(c, user) {
		return
	}

	var providers int64
	db.Model(&database.Provider{}).Count(&providers)

	var activeKeys int64
	db.Model(&database.ApiKey{}).Where("status = ?", "active").Count(&activeKeys)

	var totalKeys int64
	db.Model(&database.ApiKey{}).Count(&totalKeys)

	var activeProxies int64
	db.Model(&database.Proxy{}).Where("enabled = ?", true).Count(&activeProxies)

	var totalProxies int64
	db.Model(&database.Proxy{}).Count(&totalProxies)

	var tokens int64
	db.Model(&database.VoidToken{}).Count(&tokens)

	since := time.Now().Add(-24 * time.Hour)

	var requests24h int64
	db.Model(&database.RequestLog{}).Where("ts >= ?", since).Count(&requests24h)

	var success24h int64
	db.Model(&database.RequestLog{}).Where("ts >= ? AND success = ?", since, true).Count(&success24h)

	var failures24h int64
	db.Model(&database.RequestLog{}).Where("ts >= ? AND success = ?", since, false).Count(&failures24h)

	var tokens24hRow struct{ TotalTokens int64 }
	db.Model(&database.RequestLog{}).Where("ts >= ?", since).Select("COALESCE(SUM(total_tokens), 0) as total_tokens").Scan(&tokens24hRow)

	c.JSON(http.StatusOK, models.StatsOut{
		Providers:     int(providers),
		ActiveKeys:    int(activeKeys),
		TotalKeys:     int(totalKeys),
		ActiveProxies: int(activeProxies),
		TotalProxies:  int(totalProxies),
		Tokens:        int(tokens),
		Requests24h:   int(requests24h),
		Success24h:    int(success24h),
		Failures24h:   int(failures24h),
		Tokens24h:     int(tokens24hRow.TotalTokens),
	})
}

func RegisterStatsRoutes(router *gin.RouterGroup) {
	router.GET("/stats", getStats)
}
